docs/design/sql/05-client-support.md

```diff
@@ -0,0 +1,164 @@
+05-client-support.md
+
+# SQL Client Support
+
+## Overview
+
+This document describes the design of the client protocol and the corresponding server-side implementation.    
+
+The rest of this document is organized as follows. In section 1 we describe the requirements to the implementation. In 
+section 2 we describe the protocol design and implementation concerns. In section 3 we summarize the limitations of the current
+implementation.
+
+## 1 Requirements
+
+### 1.1 Pagination
+
+The result set might be large enough so that it doesn't fit to client's memory. 
+
+Therefore, data should be returned to the client in pages.
+
+### 1.2 Message Routing
+
+The Hazelcast Mustang engine is a distributed SQL query engine. When a query is submitted for execution, the engine splits
+it into several fragments. The fragments are executed on cluster members. 
+
+Members executing one or more fragments of the given query are called **participants**. The participant that started execution 
+of the query is called **initiator**. Participants exchange data with each other. On the final execution stage participants send
+data to the initiator to build the final result set. Only the initiator can return the final result set. 
+
+Therefore, the implementation should be able to route all requests related to the given query to the query initiator. 
+
+### 1.3 Lite Members
+
+At the moment lite members cannot start queries. I.e. a lite member can never be a query initiator. Note that this limitation
+is different from other Hazelcast subsystems. Normally, if the operation cannot be executed on the lite member, it is handed 
+over to the normal member transparently. This is not the case for the SQL purely due to time constraints: we do no have enough
+time to implement it in 4.1. This limitation will be removed in future versions.
+
+Therefore, the client should never route query execute requests to the lite members in the 4.1 release. 
+
+### 1.4 Resource Cleanup
+
+When a query is started on the initiator, certain amount of system resources are allocated. The resources are released either
+when all rows are consumed by the user, or when the explicit `close` command is invoked by the user.
+
+Therefore, the implementation must support a distinct `close` command. Also, the implementation must release the resources when
+the client disconnects.
+
+## 2 Protocol
+
+The protocol contains three commands: `execute`, `fetch`, and `close`.
+
+### 2.1 Common Entities
+
+The protocol contains the following common entities that are used across different commands:
+- **Query ID** - an opaque query identifier of type `String` that is used to locate the server-side cursor on the member
```

Comment:
Can you update this doc?

Reply:
Updated the document (see the last two commits)